haystack/pipelines/base.py

```diff
@@ -81,11 +81,28 @@ def root_node(self) -> Optional[str]:
 
     @property
     def components(self) -> Dict[str, BaseComponent]:
-        return {
-            name: attributes["component"]
-            for name, attributes in self.graph.nodes.items()
+        """
+        Returns all components used by this pipeline.
+        Note that this also includes such components that are being utilized by other components only and are not being used as a pipeline node directly.
+        """
+        node_components = [
+            attributes["component"]
+            for attributes in self.graph.nodes.values()
             if not isinstance(attributes["component"], RootNode)
-        }
+        ]
+        all_components = self._find_all_components(node_components)
+        return {component.name: component for component in all_components if component.name is not None}
+
+    def _find_all_components(self, components: List[BaseComponent]) -> Set[BaseComponent]:
```

Comment:
I would change the name of this function. Something like `_discover_utilized_components` (or similar), and move the loop in the caller. Not a big deal, but I would expect a function called `_find_all_components` to ask for no parameters, not for a nearly complete list of them :smile:

Reply:
Moving the loop would make it non-recursive, but I got an idea to get rid of the required components param.